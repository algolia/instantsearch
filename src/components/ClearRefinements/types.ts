import { ClearRefinementsRendererOptions } from '../../connectors/clear-refinements/connectClearRefinements';
import {
  ClearRefinementsCSSClasses,
  ClearRefinementsTemplates,
} from '../../widgets/clear-refinements/clear-refinements';

export type ClearRefinementsProps = {
  refine: ClearRefinementsRendererOptions['refine'];
  cssClasses: ClearRefinementsCSSClasses;
  hasRefinements: ClearRefinementsRendererOptions['hasRefinements'];
  templateProps: {
    [key: string]: any;
    templates: ClearRefinementsTemplates;
  };
};
