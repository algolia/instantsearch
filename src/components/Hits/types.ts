import { SearchResults } from 'algoliasearch-helper';
import { BindEventForHits, SendEventForHits } from '../../lib/utils';
import { Hits } from '../../types';
import { HitsCSSClasses, HitsTemplates } from '../../widgets/hits/hits';

export type HitsProps = {
  cssClasses: HitsCSSClasses;
  hits: Hits;
  results: SearchResults;
  sendEvent: SendEventForHits;
  bindEvent: BindEventForHits;
  templateProps: {
    [key: string]: any;
    templates: HitsTemplates;
  };
};
