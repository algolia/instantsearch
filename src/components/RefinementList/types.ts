export type RefinementListProps = {
  facetValues: Array<{
    isRefined: boolean;
    label: string;
    value: string;
  }>;
};
