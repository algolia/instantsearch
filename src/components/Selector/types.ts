export type SelectorProps = {
  cssClasses?: {
    root?: string | string[];
    select?: string | string[];
    option?: string | string[];
  };
  currentValue?: string | number;
  options: Array<{
    value?: string | number;
    label: string;
  }>;
  setValue(value: string): void;
};
