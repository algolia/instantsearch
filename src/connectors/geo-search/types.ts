import { SendEventForHits } from '../../lib/utils';
import {
  Connector,
  GeoLoc,
  Hit,
  TransformItems,
  WidgetRenderState,
} from '../../types';

export type GeoHit = Hit & Required<Pick<Hit, '_geoLoc'>>;

type Bounds = {
  northEast: GeoLoc;
  southWest: GeoLoc;
};

export type GeoSearchRenderState = {
  currentRefinement?: Bounds;
  position?: GeoLoc;
  items: GeoHit[];
  refine(bounds: Bounds): void;
  clearMapRefinement(): void;
  hasMapMoveSinceLastRefine(): boolean;
  isRefineOnMapMove(): boolean;
  isRefinedWithMap(): boolean;
  setMapMoveSinceLastRefine(): void;
  toggleRefineOnMapMove(): void;
  sendEvent: SendEventForHits;
};

export type GeoSearchConnectorParams = {
  enableRefineOnMapMove: boolean;
  transformItems: TransformItems<GeoHit>;
};

export type GeoSearchWidgetDescription = {
  $$type: 'ais.geoSearch';
  renderState: GeoSearchRenderState;
  indexRenderState: {
    geoSearch: WidgetRenderState<
      GeoSearchRenderState,
      GeoSearchConnectorParams
    >;
  };
  indexUiState: {
    geoSearch: {
      /**
       * The rectangular area in geo coordinates.
       * The rectangle is defined by two diagonally opposite points, hence by 4 floats separated by commas.
       *
       * @example '47.3165,4.9665,47.3424,5.0201'
       */
      boundingBox: string;
    };
  };
};

export type GeoSearchConnector = Connector<
  GeoSearchWidgetDescription,
  GeoSearchConnectorParams
>;
