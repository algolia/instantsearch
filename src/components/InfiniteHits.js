import PropTypes from 'prop-types';
import React, { Component } from 'preact-compat';
import Hits from './Hits';

class InfiniteHits extends Component {
  render() {
    const {
      cssClasses,
      hits,
      results,
      showMore,
      showMoreLabel,
      templateProps,
    } = this.props;
    const btn = this.props.isLastPage ? (
      <button disabled>{showMoreLabel}</button>
    ) : (
      <button onClick={showMore}>{showMoreLabel}</button>
    );

    return (
      <div>
        <Hits
          cssClasses={cssClasses}
          hits={hits}
          results={results}
          templateProps={templateProps}
        />
        <div className={cssClasses.showmore}>{btn}</div>
      </div>
    );
  }
}

InfiniteHits.propTypes = {
  cssClasses: PropTypes.shape({
    root: PropTypes.string,
    item: PropTypes.string,
    allItems: PropTypes.string,
    empty: PropTypes.string,
    showmore: PropTypes.string,
  }),
  hits: PropTypes.array,
  results: PropTypes.object,
  showMore: PropTypes.func,
  showMoreLabel: PropTypes.string,
  templateProps: PropTypes.object.isRequired,
  isLastPage: PropTypes.bool.isRequired,
};

export default InfiniteHits;
