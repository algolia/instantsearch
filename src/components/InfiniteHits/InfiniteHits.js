import React from 'preact-compat';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Template from '../Template/Template';

const InfiniteHits = ({
  results,
  hits,
  showMore,
  isLastPage,
  cssClasses,
  templateProps,
}) => {
  if (results.hits.length === 0) {
    return (
      <Template
        {...templateProps}
        templateKey="empty"
        rootProps={{
          className: cx(cssClasses.root, cssClasses.emptyRoot),
        }}
        data={results}
      />
    );
  }

  return (
    <div className={cssClasses.root}>
      <ol className={cssClasses.list}>
        {hits.map((hit, position) => (
          <Template
            {...templateProps}
            templateKey="item"
            rootTagName="li"
            rootProps={{ className: cssClasses.item }}
            key={hit.objectID}
            data={{
              ...hit,
              __hitIndex: position,
            }}
          />
        ))}
      </ol>

      <Template
        {...templateProps}
        templateKey="showMore"
        rootTagName="button"
        rootProps={{
          className: cx(cssClasses.loadMore, {
            [cssClasses.disabledLoadMore]: isLastPage,
          }),
          disabled: isLastPage,
          onClick: showMore,
        }}
      />
    </div>
  );
};

InfiniteHits.propTypes = {
  cssClasses: PropTypes.shape({
    root: PropTypes.string.isRequired,
    emptyRoot: PropTypes.string.isRequired,
    list: PropTypes.string.isRequired,
    item: PropTypes.string.isRequired,
    loadMore: PropTypes.string.isRequired,
    disabledLoadMore: PropTypes.string.isRequired,
  }).isRequired,
  hits: PropTypes.array.isRequired,
  results: PropTypes.object.isRequired,
  showMore: PropTypes.func,
  templateProps: PropTypes.object.isRequired,
  isLastPage: PropTypes.bool.isRequired,
};

export default InfiniteHits;
