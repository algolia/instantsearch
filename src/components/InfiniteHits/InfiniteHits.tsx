import React from 'preact-compat';
import cx from 'classnames';
import Template from '../Template/Template';
import { SearchResults } from 'algoliasearch-helper';
import { Hits } from '../../types';
import { InfiniteHitsTemplates } from '../../widgets/infinite-hits/infinite-hits';

type InfiniteHitsCSSClasses = {
  root: string;
  emptyRoot: string;
  list: string;
  item: string;
  loadPrevious: string;
  disabledLoadPrevious: string;
  loadMore: string;
  disabledLoadMore: string;
};

type InfiniteHitsProps = {
  cssClasses: InfiniteHitsCSSClasses;
  hits: Hits;
  results: SearchResults;
  hasShowPrevious: boolean;
  showPrevious: () => void;
  showMore: () => void;
  templateProps: {
    [key: string]: any;
    templates: InfiniteHitsTemplates;
  };
  isFirstPage: boolean;
  isLastPage: boolean;
};

const InfiniteHits = ({
  results,
  hits,
  hasShowPrevious,
  showPrevious,
  showMore,
  isFirstPage,
  isLastPage,
  cssClasses,
  templateProps,
}: InfiniteHitsProps): React.ReactNode => {
  if (results.hits.length === 0) {
    return (
      <Template
        {...templateProps}
        templateKey="empty"
        rootProps={{
          className: cx(cssClasses.root, cssClasses.emptyRoot),
        }}
        data={results}
      />
    );
  }

  return (
    <div className={cssClasses.root}>
      {hasShowPrevious && (
        <Template
          {...templateProps}
          templateKey="showPreviousText"
          rootTagName="button"
          rootProps={{
            className: cx(cssClasses.loadPrevious, {
              [cssClasses.disabledLoadPrevious]: isFirstPage,
            }),
            disabled: isFirstPage,
            onClick: showPrevious,
          }}
        />
      )}

      <ol className={cssClasses.list}>
        {hits.map((hit, position) => (
          <Template
            {...templateProps}
            templateKey="item"
            rootTagName="li"
            rootProps={{ className: cssClasses.item }}
            key={hit.objectID}
            data={{
              ...hit,
              __hitIndex: position,
            }}
          />
        ))}
      </ol>

      <Template
        {...templateProps}
        templateKey="showMoreText"
        rootTagName="button"
        rootProps={{
          className: cx(cssClasses.loadMore, {
            [cssClasses.disabledLoadMore]: isLastPage,
          }),
          disabled: isLastPage,
          onClick: showMore,
        }}
      />
    </div>
  );
};

export default InfiniteHits;
