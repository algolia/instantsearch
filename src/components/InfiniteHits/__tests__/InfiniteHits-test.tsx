import React from 'react';
import { mount } from 'enzyme';
import { SearchResults } from 'algoliasearch-helper';
import InfiniteHits from '../InfiniteHits';
import { Hits } from '../../../types';

describe('InfiniteHits', () => {
  const cssClasses = {
    root: 'root',
    emptyRoot: 'emptyRoot',
    item: 'item',
    list: 'list',
    loadPrevious: 'loadPrevious',
    disabledLoadPrevious: 'disabledLoadPrevious',
    loadMore: 'loadMore',
    disabledLoadMore: 'disabledLoadMore',
  };

  describe('markup', () => {
    it('should render <InfiniteHits /> on first page', () => {
      const hits: Hits = [
        {
          objectID: 'one',
          foo: 'bar',
          __position: 0,
        },
        {
          objectID: 'two',
          foo: 'baz',
          __position: 1,
        },
      ];

      const props = {
        hasShowPrevious: false,
        showPrevious: () => {},
        showMore: () => {},
        results: { hits } as SearchResults,
        hits,
        isFirstPage: true,
        isLastPage: false,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const tree = mount(<InfiniteHits {...props} />);

      expect(tree).toMatchSnapshot();
    });

    it('should render <InfiniteHits /> on last page', () => {
      const hits: Hits = [
        {
          objectID: 'one',
          foo: 'bar',
          __position: 0,
        },
        {
          objectID: 'two',
          foo: 'baz',
          __position: 1,
        },
      ];

      const props = {
        hasShowPrevious: false,
        showPrevious: () => {},
        showMore: () => {},
        results: { hits } as SearchResults,
        hits,
        isFirstPage: false,
        isLastPage: true,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const tree = mount(<InfiniteHits {...props} />);

      expect(tree).toMatchSnapshot();
    });

    it('should render <InfiniteHits /> without hits on first page', () => {
      const hits: Hits = [];

      const props = {
        hasShowPrevious: false,
        showPrevious: () => {},
        showMore: () => {},
        results: { hits } as SearchResults,
        hits,
        isFirstPage: true,
        isLastPage: false,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const tree = mount(<InfiniteHits {...props} />);

      expect(tree).toMatchSnapshot();
    });

    it('should render <InfiniteHits /> without hits on last page', () => {
      const hits: Hits = [];

      const props = {
        hasShowPrevious: false,
        showPrevious: () => {},
        showMore: () => {},
        results: { hits } as SearchResults,
        hits,
        isFirstPage: false,
        isLastPage: true,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const tree = mount(<InfiniteHits {...props} />);

      expect(tree).toMatchSnapshot();
    });

    it('should render <InfiniteHits /> with "Show previous" button on first page', () => {
      const hits: Hits = [
        {
          objectID: 'one',
          foo: 'bar',
          __position: 0,
        },
        {
          objectID: 'two',
          foo: 'baz',
          __position: 1,
        },
      ];

      const props = {
        hasShowPrevious: true,
        showPrevious: () => {},
        showMore: () => {},
        results: { hits } as SearchResults,
        hits,
        isFirstPage: true,
        isLastPage: false,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const tree = mount(<InfiniteHits {...props} />);

      const previousButton = tree.find('.loadPrevious');

      expect(previousButton.exists()).toEqual(true);
      expect(previousButton.hasClass('disabledLoadPrevious')).toEqual(true);
      expect(previousButton.props().disabled).toEqual(true);
      expect(tree).toMatchSnapshot();
    });

    it('should render <InfiniteHits /> with "Show previous" button on last page', () => {
      const hits: Hits = [
        {
          objectID: 'one',
          foo: 'bar',
          __position: 0,
        },
        {
          objectID: 'two',
          foo: 'baz',
          __position: 1,
        },
      ];

      const props = {
        hasShowPrevious: true,
        showPrevious: () => {},
        showMore: () => {},
        results: { hits } as SearchResults,
        hits,
        isFirstPage: false,
        isLastPage: true,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const tree = mount(<InfiniteHits {...props} />);

      const previousButton = tree.find('.loadPrevious');

      expect(previousButton.exists()).toEqual(true);
      expect(previousButton.hasClass('disabledLoadPrevious')).toEqual(false);
      expect(previousButton.props().disabled).toEqual(false);
      expect(tree).toMatchSnapshot();
    });
  });
});
