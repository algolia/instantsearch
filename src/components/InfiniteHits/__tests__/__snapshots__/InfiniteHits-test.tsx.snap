// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteHits markup should render <InfiniteHits /> on first page 1`] = `
<div>
  <div
    class="root"
  >
    <ol
      class="list"
    >
      <li
        class="item"
      >
        item
      </li>
      <li
        class="item"
      >
        item
      </li>
    </ol>
    <button
      class="loadMore"
    >
      showMoreText
    </button>
  </div>
</div>
`;

exports[`InfiniteHits markup should render <InfiniteHits /> on last page 1`] = `
<div>
  <div
    class="root"
  >
    <ol
      class="list"
    >
      <li
        class="item"
      >
        item
      </li>
      <li
        class="item"
      >
        item
      </li>
    </ol>
    <button
      class="loadMore disabledLoadMore"
      disabled=""
    >
      showMoreText
    </button>
  </div>
</div>
`;

exports[`InfiniteHits markup should render <InfiniteHits /> with "Show previous" button on first page 1`] = `
<div>
  <div
    class="root"
  >
    <button
      class="loadPrevious disabledLoadPrevious"
      disabled=""
    >
      showPreviousText
    </button>
    <ol
      class="list"
    >
      <li
        class="item"
      >
        item
      </li>
      <li
        class="item"
      >
        item
      </li>
    </ol>
    <button
      class="loadMore"
    >
      showMoreText
    </button>
  </div>
</div>
`;

exports[`InfiniteHits markup should render <InfiniteHits /> with "Show previous" button on last page 1`] = `
<div>
  <div
    class="root"
  >
    <button
      class="loadPrevious"
    >
      showPreviousText
    </button>
    <ol
      class="list"
    >
      <li
        class="item"
      >
        item
      </li>
      <li
        class="item"
      >
        item
      </li>
    </ol>
    <button
      class="loadMore disabledLoadMore"
      disabled=""
    >
      showMoreText
    </button>
  </div>
</div>
`;

exports[`InfiniteHits markup should render <InfiniteHits /> without hits on first page 1`] = `
<div>
  <div
    class="root emptyRoot"
  >
    empty
  </div>
</div>
`;

exports[`InfiniteHits markup should render <InfiniteHits /> without hits on last page 1`] = `
<div>
  <div
    class="root emptyRoot"
  >
    empty
  </div>
</div>
`;
